storage/feed.go

```diff
@@ -0,0 +1,273 @@
+// the split.
+type SplitRangeEvent struct {
+	Original UpdateRangeEvent
+	New      NewRangeEvent
+}
+
+// MergeRangeEvent occurs whenever a range is merged into another. This Event
+// contains two component event: an UpdateRangeEvent for the range which
+// absorbed the other, and a RemoveRangeEvent for the range that was absorbed.
+type MergeRangeEvent struct {
+	Merged  UpdateRangeEvent
+	Removed RemoveRangeEvent
+}
+
+// BeginScanRangesEvent occurs when the store is about to scan over all ranges.
+// During such a scan, each existing range will be published to the feed as a
+// NewRangeEvent. This is used because downstream consumers may be tracking
+// statistics via the Diffs in UpdateRangeEvent; this event informs subscribers
+// to clear currently cached values.
+type BeginScanRangesEvent struct{}
+
+// EndScanRangesEvent occurs when the store has finished scanning all ranges.
+// Every BeginScanRangeEvent will eventually be followed by an
+// EndScanRangeEvent.
+type EndScanRangesEvent struct{}
+
+// StoreEventFeed is a feed of events that occur on a Store. Most of these
+// events are specific to a single range within the store.
+type StoreEventFeed struct {
+	baseStoreEventPublisher
+	f *util.Feed
+}
+
+// NewStoreEventFeed creates a new StoreEventFeed. Events can immediately be
+// published and Subscribers can immediately subscribe to the feed.
+func NewStoreEventFeed(stopper *util.Stopper) StoreEventFeed {
+	feed := util.StartFeed(stopper)
+	return StoreEventFeed{
+		f: feed,
+		baseStoreEventPublisher: baseStoreEventPublisher{
+			publish: feed.Publish,
+		},
+	}
+}
+
+// Subscribe returns a util.Subscription which receives events from this
+// StoreEventFeed.
+//
+// Events are received from the feed as an empty interface and must be cast to
+// one of the event types in this package. This is best accomplished with a type
+// switch:
+//
+// 		sub := storeFeed.Subscribe()
+//		for event := range sub.Events {
+//			switch event := event.(type) {
+//			case NewRangeEvent:
+//				// Process NewRangeEvent...
+//		    case UpdateRangeEvent:
+//				// Process UpdateRangeEvent...
+//			// ... other interesting types
+//			}
+//		}
+func (sef *StoreEventFeed) Subscribe() *util.Subscription {
+	return sef.f.Subscribe()
+}
+
+// newBatch returns a new storeEventBatch which accumulates potential store
+// events as part of a batch.
+func (sef *StoreEventFeed) newBatch() *storeEventBatch {
+	seb := &storeEventBatch{
+		f:                 sef.f,
+		accumulatedEvents: make([]interface{}, 0, 3),
+	}
+	seb.baseStoreEventPublisher.publish = seb.accumulate
+	return seb
+}
+
+// storeEventBatch is used to accumulate potential Store events without
+// immediately publishing them to a feed.  This is intended for use in
+// operations that produce events but may be rolled back by an error.
+// Accumulated events will not be published to the actual feed until the
+// commit() method is called.
+type storeEventBatch struct {
```

Comment:
I created this in anticipation of needing it to handle split and merge cases; however, after discovering that a batch failure should be a fatal error, I believe it may not be necessary.

That said, I would like to leave it in for now, in case I discover I still need it when hooking it up.